File: lima_linguisticdata/scripts/transcodeIdiomatic.pl
#!/usr/bin/perl -s

# Copyright 2002-2019 CEA LIST
# SPDX-FileCopyrightText: 2022 CEA LIST <gael.de-chalendar@cea.fr>
#
# SPDX-License-Identifier: MIT

# Help mode
if ($main::h || $main::help) {
    print <<EOF;
transcode symbolic codes to numeric
usage transcode.pl /path/to/convertSymbolicCodes code-xxx.xml symbolicCode-xxx.xml rules
EOF
    exit;
}

use strict;

my @args;
my @remaining_args;
foreach my $arg (@ARGV) {
    if ($arg =~ /^\-\-/) {
        push @args, $arg;
    } else {
        push @remaining_args, $arg;
    }
}

my $argline = join(" ", @args);

die "need convertSymbolicCodes path and two files (see -help)" unless (@remaining_args > 2);

my %conv;
my $convertSymbolicCodes = shift @remaining_args;
my $codeFile = shift @remaining_args;
my $rulesFile = pop @remaining_args;

&getCodeConversion($codeFile, @remaining_args);

my @conv_keys = keys %conv;
if ($#conv_keys < 0) {
    die "ERROR: transcodeIdiomatic.pl: can't load conv keys\n";
}

open(FIN,$rulesFile) || die "cannot open $rulesFile";
while (<FIN>) {
    #if (/(.*):(ABS_)?IDIOM\$([^:]*)(.*)/) {
    # category can contain a ":" but normalized form should not
    if (/(.*):(ABS_)?IDIOM\$(.*)(:.*)/) {
        if (! exists($conv{$3})) {
            #print STDERR "transcodeIdiomatic.pl: can't find conv for \"$3\"\n";
        }
        print "$1:$2IDIOM\$".$conv{$3}.$4."\n";
    } else {
        print;
    }
}
close(FIN);

sub getCodeConversion {
    my ($codeFile, @symbCodeFiles) = @_;
    if ( !open(FCONV,"$convertSymbolicCodes $argline --code=$codeFile " .  join(" ", @symbCodeFiles) . " |") ) {
        print STDERR "cannot get code conversion: $!\n";
        exit(1);
    }

    while(<FCONV>) {
        /(.*);(.*);/;
        $conv{$1} = $2;
    }

    close(FCONV);
}

File: lima_linguisticdata/scripts/transcode_idiomatic.py
import argparse
import re
import subprocess
import sys

from tqdm import tqdm


def getCodeConversion(convertSymbolicCodes, codeFile, symbolicCodeFiles):
    conv = {}
    command = [convertSymbolicCodes, '--code=' + codeFile] + symbolicCodeFiles
    try:
        result = subprocess.run(command, stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE, text=True, check=True)
    except subprocess.CalledProcessError as e:
        sys.stderr.write(f"cannot get code conversion: {e}\n")
        sys.exit(1)

    for line in result.stdout.splitlines():
        parts = line.split(';')
        if len(parts) >= 2:
            conv[parts[0]] = parts[1]

    return conv


def main():
    parser = argparse.ArgumentParser(
        description='Transcode symbolic codes to numeric codes.')
    parser.add_argument('convertSymbolicCodes',
                        help='Path to the convertSymbolicCodes script')
    parser.add_argument('codeFile', help='Code file')
    parser.add_argument('symbolicCodeFiles', nargs='+',
                        help='Symbolic code files')
    parser.add_argument('rulesFile', help='Rules file')

    args = parser.parse_args()

    convertSymbolicCodes = args.convertSymbolicCodes
    codeFile = args.codeFile
    rulesFile = args.rulesFile
    symbolicCodeFiles = args.symbolicCodeFiles

    conv = getCodeConversion(convertSymbolicCodes, codeFile, symbolicCodeFiles)

    if not conv:
        sys.stderr.write(
            "ERROR: transcodeIdiomatic.py: can't load conv keys\n")
        sys.exit(1)

    with open(rulesFile, 'r') as fin:
        lines = fin.readlines()

    with tqdm(total=len(lines), desc="Transcoding rules") as pbar:
        for line in lines:
            pbar.update(1)
            match = re.match(r'(.*):(ABS_)?IDIOM\$(.*)(:.*)', line)
            if match:
                prefix, abs_idiom, key, suffix = match.groups()
                if key not in conv:
                    # Uncomment the next line if you want to print missing
                    # keys warnings
                    # sys.stderr.write(
                    # f"transcodeIdiomatic.py:
                    # can't find conv for \"{key}\"\n")
                    continue
                print(f"{prefix}:{abs_idiom or ''}IDIOM${conv[key]}{suffix}")
            else:
                print(line, end='')


if __name__ == "__main__":
    main()
